caddyhttp/proxy/upstream.go

```diff
@@ -416,9 +423,21 @@ func (u *staticUpstream) healthCheck() {
 			}
 
 			if r, err := u.HealthCheck.Client.Do(req); err == nil {
-				io.Copy(ioutil.Discard, r.Body)
-				r.Body.Close()
-				unhealthy = r.StatusCode < 200 || r.StatusCode >= 400
+				if u.HealthCheck.ContentString != "" {
+					b, err := ioutil.ReadAll(r.Body)
```

Comment:
I've been looking at this line for a few minutes, staring... glaring... because usually doing a ReadAll() is inefficient for memory pressure. However, health check endpoints are purely static (hard-coded by the site owner), right? They're trusted endpoints that (presumably) won't have a large response body. If these response bodies are just a few (< ~1K?) bytes on average, I think this is fine as-is.

However, a better design here would be to implement a reader that wraps r.Body and checks the contents of every `Read()`, advancing a pointer/index into the `ContentString` value, resetting it when there's a mismatch, and discarding all the reads after the search finds what it's looking for (`idx == len(ContentString)-1`). This way, we don't need any additional buffers. (This sparked quite the intriguing discussion on Gophers Slack! Reminds me of an interview question.)

Do you think such an optimization will be necessary/useful?